### DISCUSSION TITLE
Task 2 results

### DISCUSSION THREAD

**Discussion by erlik-han** [cafawo/FinancialDataAnalytics#8] [Q&A] [answered]
Hello I have one question about task 2. I am getting this results 
Price using numerical integration:  20.93142032537204
Price using Monte Carlo simulation:  89.31532758188021

They are really big differences between two results.Is it normal or something wrong with my code?

Thanks a lot for your contributions

**Comment by erlik-han** [comment]
this is my code:

```
###TASK2
import numpy as np
from scipy.integrate import quad
from scipy.stats import norm

# Set parameters
S0 = 220      # Today's price of the underlying
K = 220       # Strike price
sigma = 0.98  # Volatility
r = 0.1       # Risk-free interest rate
T = 1         # Maturity (in years)

# Calculate parameters for log-normal distribution
mu = np.log(S0) + (r - 0.5 * sigma ** 2) * T
s2 = sigma ** 2 * T

# Define integrand function for numerical integration
def integrand(x):
    val = (np.exp(x) - K) * norm.pdf(x, mu, s2**0.5)
    return val

# Calculate price using numerical integration
price_int, _ = quad(integrand, -10, 10)

if np.isnan(price_int):
    print("Error: Price using numerical integration is NaN.")
else:
    price_int *= np.exp(-r * T)
    print("Price using numerical integration: ", price_int)

# Calculate price using Monte Carlo simulation
N = 1000000   # Number of simulations
ST = S0 * np.exp((r - 0.5 * sigma ** 2) * T + sigma * np.sqrt(T) * np.random.normal(size=N))

price_mc = np.mean(np.maximum(ST - K, 0)) * np.exp(-r * T)

print("Price using Monte Carlo simulation: ", price_mc)
```

**Reply by cafawo** [reply]
Please use code blocks to make your code more readable: https://www.markdownguide.org/extended-syntax/

**Reply by erlik-han** [reply]
Okay, thanks.

**Comment by erlik-han** [comment]
Also I have another attemp and result is this 
Price using numerical integration:  nan
Price using Monte Carlo simulation:  89.60167975820947

the code is below:

```
import numpy as np
from scipy.integrate import quad
from scipy.stats import norm

# Set parameters
S0 = 220      # Today's price of the underlying
K = 220       # Strike price
sigma = 0.98  # Volatility
r = 0.1       # Risk-free interest rate
T = 1         # Maturity (in years)

# Calculate parameters for log-normal distribution
mu = np.log(S0) + (r - 0.5 * sigma ** 2) * T
s2 = sigma ** 2 * T

# Define integrand function for numerical integration
def integrand(x):
    val = (np.exp(x) - K) * norm.pdf(x, mu, s2**0.5)
    return val

# Calculate price using numerical integration
price_int, _ = quad(integrand, -np.inf, np.inf)
price_int *= np.exp(-r * T)

print("Price using numerical integration: ", price_int)

# Calculate price using Monte Carlo simulation
N = 1000000   # Number of simulations
ST = S0 * np.exp((r - 0.5 * sigma ** 2) * T + sigma * np.sqrt(T) * np.random.normal(size=N))

price_mc = np.mean(np.maximum(ST - K, 0)) * np.exp(-r * T)

print("Price using Monte Carlo simulation: ", price_mc)
```

**Reply by Di3goCH** [reply, marked as answer]
Hi, I have the same result and almost identical code in the Monte Carlo simulation. However, in the integration part, we have some differences in the code, but my result is also lower than the one from Monte Carlo simulation. 

I wanted to ask you:

1) why do you use norm.pdf() and not lognormal distribution?  "mu" that we calculate seems to be for the lognormal distribution.

2) maybe the integration should be from K to +inf because the maximization function does not admit negative values from (x-K).

**Reply by cafawo** [reply]
Correct, $S_T$ follows a Log-Normal distribution (relevant for the integration task). 

Note that $\log S_T / S_0$ follows a Normal distribution (relevant for MC).

The Call option is exercised if $S_T > K$, hence, $K$ to $+ \inf$.

**Reply by erlik-han** [reply]
1-Actually I was trying different codes to get results also I have this alternative and they are giving the same results 

```
# Define integrand function for numerical integration
def integrand(x):
    val = (np.exp(x) + K) * np.exp(-0.5 * ((x - mu) / s) ** 2) / (s * np.sqrt(2 * np.pi)) #I cahnged the -K to +K
    return val
```

2- As you suggest that I changed x-K to x+K for both of them and I got very similar results for both of them .It looks like the problem was solved for me.
Price using numerical integration:  419.0598123566462
Price using Monte Carlo simulation:  419.0889277297257
